### PR TITLE
feat: make flow distributed work&tests

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
preq: https://github.com/GreptimeTeam/greptimedb/pull/4254
## What's changed and what's your intention?

make flownode dist start work

- add `FrontendClient` for flownode to connect to frontend for writing back results
- add sqlness tests for dist flownode
- modify sqlness runner to start flownode

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Added comprehensive configuration settings for Flownode, including modes, gRPC options, heartbeat intervals, logging, and tracing.
  - Introduced a frontend client for database operations in the flow module.
  - Implemented compression encoding for gRPC communication in FlownodeServer.

- **Enhancements**
  - Updated error handling to support new TonicTransport variant.
  - Improved distributed mode configuration with new settings and defaults.

- **Documentation**
  - Reorganized configuration documentation, adding a new section for Flownode settings.

- **Tests**
  - Enhanced test environment to support Flownode server process.

### CODE REVIEW

File: src/flow/src/fe_client.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+fn to_rpc_request(request: Request, ctx: &QueryContext) -> GreptimeRequest {
+    let header = RequestHeader {
+        catalog: ctx.current_catalog().to_string(),
+        schema: ctx.current_schema().to_string(),
+        authorization: None,
+        // dbname is empty so that header use catalog+schema to determine the database
+        // see `create_query_context` in `greptime_handler.rs`
+        dbname: "".to_string(),
+        timezone: ctx.timezone().to_string(),
+        tracing_context: W3cTrace::new(),
```
Comment (WenyXu) [inline src/flow/src/fe_client.rs:54]: How about to keep the tracing context
```suggestion
        tracing_context: TracingContext::from_current_span().to_w3c(),
```

File: src/cmd/src/flownode.rs [hunk 2]
```diff
@@ -119,16 +120,26 @@ impl SubCommand {
...
     #[clap(long)]
     rpc_addr: Option<String>,
+    /// Hostname for the gRPC server.
     #[clap(long)]
     rpc_hostname: Option<String>,
+    /// Metasrv address list;
     #[clap(long, value_delimiter = ',', num_args = 1..)]
     metasrv_addrs: Option<Vec<String>>,
+    /// The gprc address of the frontend server used for writing results back to the database.
+    /// Need prefix i.e. "http://"
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/cmd/src/flownode.rs:136]: > Why need the prefix?

tonic::Endpoint (https://docs.rs/tonic/latest/tonic/transport/channel/struct.Endpoint.html#method.from_shared) need prefix to connect as a client so

File: config/flownode.example.toml [hunk 3]
```diff
@@ -0,0 +1,93 @@
+## The running mode of the flownode. It can be `standalone` or `distributed`.
```
Comment (zyy17) [inline config/flownode.example.toml:1]: You can put the toml in `config-docs-template.md` and executes the `make config-docs`.

File: src/cmd/src/flownode.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/cmd/src/flownode.rs:136]: Should we support to `client_tls_config`? https://github.com/GreptimeTeam/greptimedb/blob/b5c6c72b025b569d7333c23f65090bbe33d08d62/src/common/grpc/src/channel_manager.rs#L165

File: src/flow/src/adapter.rs [hunk 4]
```diff
@@ -97,6 +98,7 @@ impl Default for FlownodeOptions {
             cluster_id: None,
             node_id: None,
             grpc: GrpcOptions::default().with_addr("127.0.0.1:3004"),
+            frontend_addr: None,
```
Comment (zyy17) [inline src/flow/src/adapter.rs:101]: question: what if there are multiple frontend instances? Should we use a list of frontend addr?

Comment (discord9) [reply src/flow/src/adapter.rs:101]: > question: what if there are multiple frontend instances? Should we use a list of frontend addr? Or it's ok to use any of the frontend instances, which means we can specify it as a load balancer addr.

Yes as long as they are grpc service

**Review by WenyXu — APPROVED** [review]

**Review by zyy17 — APPROVED** [review]

**Pull request merged**